lib/Sema/CSSyntacticElement.cpp

```diff
@@ -2159,6 +2175,38 @@ class SyntacticElementSolutionApplication
     return returnStmt;
   }
 
+  ASTNode visitThenStmt(ThenStmt *thenStmt) {
+    // Note we're defensive here over whether we're in a SingleValueStmtExpr,
+    // as we don't diagnose out-of-place ThenStmts until MiscDiagnostics. If we
+    // have an out-of-place ThenStmt, just rewrite the expr without a contextual
+    // type.
```

Comment:
Is there not enough information to determine whether a `then` is out of place during constraint generation for a syntactic element? We could fail invalid uses in regular solving and produce a fix in diagnostic mode?

Reply:
I initially wanted to keep all the logic together in the MiscDiagnostics walker, but yeah I think doing it this way is probably better, updated to use a fix.

Reply:
Thank you!